File: experiments/breakdown/base/tpch_q1.cpp
#include "../tpch_q1_util.h"

int main(int argc, char *argv[]) {
    struct timeval start, stop;
    string file = argv[1];

    char *l_returnflag = new char[Q1_LINEITEM];
    char *l_linestatus = new char[Q1_LINEITEM];
    float *l_quantity = new float[Q1_LINEITEM];
    float *l_extendedprice = new float[Q1_LINEITEM];
    float *l_discount = new float[Q1_LINEITEM];
    float *l_tax = new float[Q1_LINEITEM];
    long *l_shipdate = new long[Q1_LINEITEM];
    init(file, l_returnflag, l_linestatus, l_quantity, l_extendedprice,
            l_discount, l_tax, l_shipdate);
    unordered_map<short,char *> *vals = new unordered_map<short,char *>();

    test(vals);
    gettimeofday(&start, NULL);
    for (int i = 0; i < Q1_LINEITEM; i++)
        if (l_shipdate[i] <= Q1_DATE1) {
            short key = l_returnflag[i] << 8;
            key |= l_linestatus[i];
            auto itr = vals->find(key);
            char *val;
            if (itr == vals->end()) {
                val = new char[Q1_VALSIZE];
                val[0] = l_returnflag[i];
                val[1] = l_linestatus[i];
                *((int *) (val + 2)) = 0;
                *((float *) (val + 6)) = 0.0f;
                *((float *) (val + 10)) = 0.0f;
                *((float *) (val + 14)) = 0.0f;
                *((float *) (val + 18)) = 0.0f;
                *((float *) (val + 22)) = 0.0f;
                (*vals)[key] = val;
            }
            else
                val = itr->second;
            *((int *) (val + 2)) += 1;
            *((float *) (val + 6)) += l_quantity[i];
            *((float *) (val + 10)) += l_extendedprice[i];
            *((float *) (val + 14)) += l_extendedprice[i] * (1 - l_discount[i]);
            *((float *) (val + 18)) += l_extendedprice[i] * (1 - l_discount[i])
                                     * (1 + l_tax[i]);
            *((float *) (val + 22)) += l_discount[i];
        }
    gettimeofday(&stop, NULL);
    test(vals);
    printTime("tpch_q1", start, stop);

    delete[] l_returnflag;
    delete[] l_linestatus;
    delete[] l_quantity;
    delete[] l_extendedprice;
    delete[] l_discount;
    delete[] l_tax;
    delete[] l_shipdate;
    delete vals;

    return 0;
}

File: experiments/breakdown/java/TpchQ1.java
import java.io.BufferedReader;
import java.io.FileReader;
import java.text.SimpleDateFormat;
import java.util.HashMap;

public class TpchQ1 {
    public static final double NSEC = 0.000000001;
    public static final int LINEITEM = 59986052;
    public static final long DATE1 = 904708800;

    public static void main(String[] args) {
        char[] l_returnflag = new char[LINEITEM];
        char[] l_linestatus = new char[LINEITEM];
        float[] l_quantity = new float[LINEITEM];
        float[] l_extendedprice = new float[LINEITEM];
        float[] l_discount = new float[LINEITEM];
        float[] l_tax = new float[LINEITEM];
        long[] l_shipdate = new long[LINEITEM];

        HashMap<Short,Q1Tuple> vals = new HashMap<>();

        try (
            BufferedReader br = new BufferedReader(new FileReader(args[0]));
        ) {
            SimpleDateFormat sdf = new SimpleDateFormat("yyyy-MM-dd");
            String ln = "";
            for (int i = 0; (ln = br.readLine()) != null; i++) {
                String[] toks = ln.split("\\|");
                l_quantity[i] = Float.parseFloat(toks[4]);
                l_extendedprice[i] = Float.parseFloat(toks[5]);
                l_discount[i] = Float.parseFloat(toks[6]);
                l_tax[i] = Float.parseFloat(toks[7]);
                l_returnflag[i] = toks[8].charAt(0);
                l_linestatus[i] = toks[9].charAt(0);
                l_shipdate[i] = sdf.parse(toks[10]).getTime() / 1000;
            }
        }
        catch (Exception e) {
            e.printStackTrace();
            System.exit(1);
        }

        long start = System.nanoTime();
        for (int i = 0; i < LINEITEM; i++)
            if (l_shipdate[i] <= DATE1) {
                short key = (short) (l_returnflag[i] << 8);
                key |= l_linestatus[i];
                Q1Tuple val = vals.get(key);
                if (val == null) {
                    val = new Q1Tuple(l_returnflag[i], l_returnflag[i]);
                    vals.put(key, val);
                }
                val.r_quantity += l_quantity[i];
                val.r_extendedprice += l_extendedprice[i];
                val.r_discountprice += l_extendedprice[i] * (1 - l_discount[i]);
                val.r_discounttax += l_extendedprice[i] * (1 - l_discount[i])
                                   * (1 + l_tax[i]);
                val.r_discount += l_discount[i];
            }
        long stop = System.nanoTime();

        System.err.println(vals);
        System.err.println("tpch_q1: " + ((stop - start) * NSEC));
    }

    private static class Q1Tuple {
        public final char r_returnflag;
        public final char r_linestatus;
        public float r_quantity;
        public float r_extendedprice;
        public float r_discountprice;
        public float r_discounttax;
        public float r_discount;

        public Q1Tuple(char r_returnflag, char r_linestatus) {
            this.r_returnflag = r_returnflag;
            this.r_linestatus = r_linestatus;
            r_quantity = 0.0f;
            r_extendedprice = 0.0f;
            r_discountprice = 0.0f;
            r_discounttax = 0.0f;
            r_discount = 0.0f;
        }

        public String toString() {
            return "(" + r_returnflag + ","
                       + r_linestatus + ","
                       + r_quantity + ","
                       + r_extendedprice + ","
                       + r_discountprice + ","
                       + r_discounttax + ","
                       + r_discount + ")";
        }
    }
}
